### PR TITLE
add TS types for partial config options

### LINKED ISSUE #7238: Cannot provide types cypress.run config options
Cypress v4.5.0

When writing TypeScript `cypress.run(...)` code using NPM module API, cannot type the config object

```ts
const config: Cypress.ConfigOptions = {

}
cypress.run({ config })
```

Screen Shot 2020-05-06 at 11 11 15 AM

Trying to use `Partial` in user-space still leads to problem because we have type duplication and divergence

```ts
const config: Partial<Cypress.ConfigOptions> = {

}
cypress.run({ config })
```

Screen Shot 2020-05-06 at 11 11 33 AM

### PR DESCRIPTION
**Pull request by bahmutov**
- Closes #7238

### User facing changelog

We export `Cypress.ConfigOptions` as a partial of the full options interface, thus you can write realistic TypeScript code right away

```ts
// provide only some config options
const runConfig: Cypress.ConfigOptions = {
  baseUrl: 'http://localhost:8080',
  env: {
    login: false
  },
}
cypress.run({ config: runConfig })
```

### Additional details

- removed duplicate TS code between `npm-api ts` file and `index.d.ts`
- moved the full config options into a different interface name, and added
```
type ConfigOptions = Partial<AllConfigOptions>
```

### How has the user experience changed?

Using Cypress NPM module API correctly got easier

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?

### CODE REVIEW

**Comment by bahmutov** [comment]
I think this is a better solution than https://github.com/cypress-io/cypress/pull/6696

**Review by flotwig — CHANGES_REQUESTED** [review]
so the reason for this, over #6696, is that it allows us to express the shape of the config *after* config.coffee has resolved the full configuration object, right?

File: cli/types/index.d.ts [hunk 1]
```diff
@@ -2287,7 +2287,7 @@ declare namespace Cypress {
     multiple: boolean
   }
 
-  interface ConfigOptions {
+  interface AllConfigOptions {
```
[the remark this answers is not in the document]
Comment (bahmutov) [reply cli/types/index.d.ts:2290]: I will use `ResolvedConfigOptions` - just note that in our server package the resolved config has options AND for each has `from: ... value: ...` structure

**Comment by bahmutov** [comment]
unrelated flake e2e test

File: cli/types/index.d.ts [hunk 2]
```diff
@@ -2458,7 +2458,13 @@ declare namespace Cypress {
     experimentalGetCookiesSameSite: boolean
   }
 
-  interface PluginConfigOptions extends ConfigOptions {
+  /**
+   * When users specify config options, they can provide only
+   * some properties.
+   */
```
Comment (flotwig) [inline cli/types/index.d.ts:2464]: The way that this is phrased implies that users cannot supply all config options, maybe something like this:

```suggestion
  /**
   * All configuration items are optional.
   */
```

**Review by flotwig — CHANGES_REQUESTED** [review]

**Review by flotwig — APPROVED** [review]

**Pull request merged**